aptos-move/block-executor/src/executor.rs

```diff
@@ -305,6 +349,17 @@ where
                 .expect("Read set must be recorded");
             let mut execution_still_valid =
                 read_set.validate_delayed_field_reads(versioned_cache.aggregators(), txn_idx);
+
+            match last_input_output.output_category(txn_idx) {
+                Some(ErrorCategory::SpeculativeExecutionError) => {
+                    assert!(!execution_still_valid);
```

Comment:
This assert seems scary, I think we are asserting that only because of a read we could have had speculative execution error that's not re-executed (seems fair based on our discussion, e.g. in other cases like delta history problems, we would re-execute*), and that the flag in CapturedReads will be set? (because the values can change again to pass validation). Seems easy to have an issue, can we just add a comment and fall back to CodeInvariantError in this case?

*: Also seems like we would have to separate the aggregated reads speculative_error in Captured Reads from the other one (e.g. metadata/exists/value checks), because in the former we don't want to re-execute the txn, while in the later we do?